### ISSUE TITLE
should some type arguments not be affected by surrounding defaults?

### ISSUE THREAD

**Issue by kevin1e100** [jspecify/jspecify#46]
Consider a class like Guava's `ImmutableList`, whose type parameter would presumably be declared as `class ImmutableList<T extends @NotNull Object>`, since the class is null-hostile.  Now in another source file:
```
class Foo {
@DefaultNullable
ImmutableList<String> names() {...}

@DefaultNullnessUnknown
ImmutableList<String> values() {...}

ImmutableList<@NullnessUnknown String> keys() {...}

@DefaultNullnessUnknown
List<String> modifiable() {...}
}
```

- Should we consider `names()` to return a `ImmutableList<@NotNull String>` despite the conflicting default?  That seems convenient but is non-obvious from looking at `Foo` alone.
- Should we expect the `@NotNull` to be made explicit, since it conflicts with the default?  That avoids the non-obvious problem from the last question but would require "spraying" essentially redundant `@NotNull` annotations.
- Should we even worry about `@DefaultNullable`, since we hope it's rarely used?  If it's rarely used then maybe requiring explicit `@NotNull` is just tolerable.

What about `values()`, which has unknown default?  I think things get extra-tricky here, since this will happen in unannotated legacy code that happens to be using `ImmutableList`, and it seems implausible, therefore, to require explicit `@NotNull` qualifiers, therefore (at least when there's no default annotation; maybe it's plausible with explicit `@DefaultNullnessUnknown` but we've so far tried not to distinguish "no default" from `@DefaultNullnessUnknown`).

It seems desirable to interpret `values()` to return `ImmutableList<@NotNull String>` regardless.  But does that match the user's intuition in the context of "unknown nullness by default"?  Possibly not but on the other hand there seems little harm at least in this example.

One counter-point could be consider an example like `keys()`.  Here the user was really clear that they want to consider the returned `ImmutableList`'s elements to be considered `@NullnessUnknown`, and it seems problematic to silently "upgrade" that to `@NotNull` elements.  On the other hand we could consider such a use of `@NullnessUnknown` to be invalid, though that would be the first time we've considered making unknown nullness invalid. In the case of `ImmutableList` that may be plausible, but I'm not sure we can make cases like `keys()` invalid in general.

One other point maybe related point is to consider `modifiable()`, assuming `List` is declared to allow nullable elements, i.e., `interface List<T extends @Nullable Object>`.  Here we presumably would interpret `modifiable()` to return effectively `List<@NullnessUnknown String>`, and given that it seems potentially confusing if `values()`, having the same annotations otherwise, effectively returns `ImmutableList<@NotNull String>`.  But again, while it feels inconsistent maybe there is just no harm.

**Comment by kevinb9n** [comment]
@kevin1e100 I think the content in here has gone stale, and has become hard to parse in our new many-fewer-annotations world.

It's likely most of the original concerns at the heart of it probably sort into one of the buckets "came up again", "will come up again", or "no longer applicable".

But if there are aspects of this discussion that are still relevant I think I would have to count on you to restate them under today's worldview, in new issues. Short of that this will probably just keep getting ignored so I'm going to go ahead and close it. If that feels wrong but you don't have time to do as requested, then just mash that reopen button and I'll do my best.
